### ISSUE TITLE
Allow tabs to wrap to multi-line

### ISSUE THREAD

**Issue by illz** [microsoft/vscode#70413]
I almost always have more tabs open than fit in the horizontal space provided. I have seen the thread on shrinking tabs to try to fit them in one line [#15048](https://github.com/Microsoft/vscode/issues/15048), and I've also seen this closed feature request for what I'm asking for - [#32836](https://github.com/Microsoft/vscode/issues/32836). That issue was closed saying there would be no changes for 6-12 months, and it has been around 15 months now so wanted to reopen discussion.

JetBrains IDEs added this a few years ago - simple config option to wrap open tabs to multiple lines rather than extending indefinitely horizontally, forcing a horizontal scroll or use of the Open Editors section in the left bar. While the Open Editors block is a nice idea, I've spent so many years switching via tabs (and continue to do so in all my other editors) that I'm requesting a better solution.

**Comment by hugodichon** [comment]
For sure that is a really basic and essential functionality to have. Following up on that for a while now.

**Comment by joaohonorato** [comment]
Can this be done any time soon, i just love this feature, it awe me to not see it in vscode.

**Comment by neil-119** [comment]
Would love to have this as well. We use VS Code mostly for Angular work and we typically have multiple components (TS, HTML, CSS) and services (TS) open at a time, and we like to group relevant tabs together at the top.

**Comment by helloli** [comment]
Although we can use 'Open Editors' or 'Ctrl + Tab' instead, many of us would appreciate it very much if multi-line tabs were supported. Please carefully consider it.

**Comment by akira90** [comment]
Has there been any updates to this feature?

**Comment by im-richard** [comment]
Seems like really simple functionality which is almost a must. I've debated utilizing other apps just for this feature alone because at any given time, I can have roughly 20-30 tabs open, and everything on a single row is just annoying at times.

Really hope this is looked at as it's literally the only pet-peeve I currently have with vscode.

**Comment by hugodichon** [comment]
Come on guys... At least provide some feedback to people expecting that feature one day.
If I had enough time I would have liked to investigate and submit a pull request, unfortunately I don't have that luxury.
I would like to add up that Microsoft main development product has that feature built-in since a bunch of year yet, it's called Visual Studio.

**Comment by akira90** [comment]
I've gave up with VSCODE - this feature should've been built in, even Eclipse has it. I'm using IntelliJ now as a lot of things come out of the box.

**Comment by lucaszanella** [comment]
PLEASE, THIS FEATURE IS EXTREMELY USEFUL

Look at my firefox: https://imgur.com/a/CKOGGMq

I can't live without miltiline tabs

It's VERY annoying to search for the opened tabs by scrolling. I'm almost giving up on vscode because of this.

**Comment by stevenlaidlaw** [comment]
I've implemented a stop-gap solution using [VSCode Custom CSS](https://github.com/be5invis/vscode-custom-css). Just create your own custom CSS file in your system and fill it with the following:

```css
.tabs-and-actions-container > .monaco-scrollable-element {
  height: auto !important;
}

.tabs-and-actions-container > .monaco-scrollable-element > .tabs-container {
  height: auto !important;
  flex-wrap: wrap;
}
```

Then just point the plugin at your custom file in your settings.json like so:
```json
    "vscode_custom_css.imports": [
        "file:///home/username/path/to/dir/custom.css"
    ],
    "vscode_custom_css.policy": true,
```

Then hit `ctrl-shift-p` and run "Reload custom CSS and JS"

This has solved the problem for me until they implement an in-app solution.

**Comment by manustays** [comment]
> ```css
> .tabs-and-actions-container > .monaco-scrollable-element > .tabs-container {
>   height: auto !important;
>   flex-wrap: wrap;
> }
> ```

@stevenlaidlaw, your solution worked for me, thanks! How do you figure out the **classes used by vscode**?

**Comment by stevenlaidlaw** [comment]
@manustays you can toggle the chrome dev tools via the action menu (ctrl+shift+P)

**Comment by akira90** [comment]
After applying the custom.css I got an error stating "Installation appears to be corrupt, please reinstall". Did any of you experience something similar ?

**Comment by stevenlaidlaw** [comment]
@akira90 As detailed in the plugin (https://github.com/be5invis/vscode-custom-css):

> If Code complains about that it is corrupted, simply click “Don't show again”

**Comment by akira90** [comment]
@stevenlaidlaw 

It keeps appearing again and again, after closing it appears again after 30 seconds. Then cmd + q is disabled as you can't close vscode because of the shortcut cannot be found, checked keymaps - it's still there. 

This issue appeared after I've installed the Custom CSS and JS extension.

**Comment by lucaszanella** [comment]
I just won't try this because it requires running VSCode as sudo and I don't like this idea

Please Microsoft, provide this by default

**Comment by Manachi** [comment]
I'm amazed and disappointed this option isn't available by default in Visual Studio Code. It exists in every other IDE or text editor in the market.  Please prioritise this feature Microsoft. 

Is there at least an extension that does this in the meantime anyone could point me to?

**Comment by kenjione** [comment]
> I'm amazed and disappointed this option isn't available by default in Visual Studio Code. It exists in every other IDE or text editor in the market. Please prioritise this feature Microsoft.
> 
> Is there at least an extension that does this in the meantime anyone could point me to?

@Manachi,

If you look up you can find @stevenlaidlaw's post how he has solved the issue. You need to install **vscode-custom-css** extension and copy the stylesheet which you can find in his (or @manustays's) comment.

**Comment by rmolinamir** [comment]
> Look at my firefox: https://imgur.com/a/CKOGGMq

It would be amazing for vscode to achieve this, was about to open an issue until I found this :)

Has there been any announcements regarding this feature?

**Comment by sebaranowski** [comment]
> I just won't try this because it requires running VSCode as sudo and I don't like this idea
Agree. A new settings field for an additional css file or raw styles string for overwrites would help too. Atom has this too and I love it - this makes the DevTools inside vscode so much more useful..

**Comment by josineiaraujo** [comment]
This is an option I wish I could use!

**Comment by rajnisheu** [comment]
Its unbelievable there is no mutli line tab support. The Open Editors list is so irritating to the eye jumping up and done when files are opened and closed.

+1 for the issue

**Comment by dhex** [comment]
This is not just nice-to-have. It is a must. Waiting.. for official news ...

**Comment by hoodist** [comment]
Multi-line-tabs is very useful feature for editors. 
'Open Editors Block' is good idea but it's not enough for us.

I hope that Microsoft adding to this feature.

**Comment by nausicaasnow** [comment]
Definitely a **must have** feature.

**Comment by iki** [comment]
combined @stevenlaidlaw and @manustays styles and added `monaco-grid-view` overflow to avoid hiding the tabs under top app bar

```css
.monaco-grid-view {
  overflow: visible;
}

/* .monaco-workbench .part.editor > .content .editor-group-container > */

.title.tabs > .tabs-and-actions-container > .monaco-scrollable-element {
  height: auto !important;
}

.title.tabs > .tabs-and-actions-container > .monaco-scrollable-element > .tabs-container {
  height: auto !important;
  flex-wrap: wrap;
}

.title.tabs > .tabs-and-actions-container > .monaco-scrollable-element > .tabs-container > .tab {
  height: 25px;
  padding-left: 4px;
  /* font-size: 0.9em; */
}

.title.tabs > .tabs-and-actions-container > .monaco-scrollable-element > .tabs-container > .tab .label-name {
  font-size: inherit !important;
}

.title.tabs > .tabs-and-actions-container > .monaco-scrollable-element > .tabs-container > .tab > .tab-close {
  width: 20px;
}

.title.tabs > .tabs-and-actions-container > .monaco-scrollable-element > .tabs-container > .tab > .tab-close .action-label {
  height: 12px;
  width: 12px;
  background-size: 12px;
}
```

**Comment by pzuraq** [comment]
So I'm probably in a very small minority here, but I find horizontal tab bars to be very difficult to work with, especially when I have lots of code open in a large codebase and am jumping around a lot 😅 

A couple of times now, I've hacked together a vertical tab bar to get around this, it works much better for me personally:

![Screen Shot 2019-11-14 at 8 57 50 PM](https://user-images.githubusercontent.com/685518/68967228-9683dd80-0794-11ea-8eb5-a0f56ba3ff57.png)

VSCode doesn't support this type of UI hacking at all (the solution is pretty terrible), and I imagine it would be a large lift to make it so extensions could add arbitrary panels and views like this, but it would also be really nice. 

One of the things that is so great about a browser based editor is that it really is pretty hackable, and it mean we can experiment with new UIs and better DX in ways that would never have been possible in something like IntelliJ or Visual Studio directly. I know I've had lots of random ideas for utilities and widgets that would make life easier, would be cool to be able to play around more with the layout and such.

**Comment by Lady-Natalya** [comment]
Come on Microsoft, please add this feature or respond to us why y'all aren't going to add it.

Horizontal scrolling the tab bar is horrible.  And the bar itself is by default hidden so I have to hover where I think it might be and await for it to appear.  And it's so thin, I tried making it thicker so it would be easier to click on, but that didn't work either.

I just don't understand, do y'all want a bad UI for this product?  Please help us.

**Comment by lucaszanella** [comment]
commenting again here: PLEASE, Microsoft, add this feature. It's a PAIN to do horizontal scrolling. Mouses weren't made for horizontal scrolling.

Enviado via ProtonMail móvel

-------- Mensagem Original --------
Ativo 15 de nov de 2019 17:23, Natalya escreveu:

**Comment by Manachi** [comment]
Microsoft, fix it.

**Comment by ciltocruz** [comment]
![image](https://user-images.githubusercontent.com/19903919/69335597-9f295780-0c5d-11ea-9b16-bc62258d483b.png)

Netbeans has the option to create a bar for each project. VScode does not have "projects" but VScode does have Folders within the workspace ... It would be nice if a bar of open files could be created for each folder in the workspace.

**Comment by ciltocruz** [comment]
![image](https://user-images.githubusercontent.com/19903919/69336344-92593380-0c5e-11ea-9176-9c6e1c2b83ad.png)

**Comment by rafaelucena** [comment]
@Lady-Natalya
> Come on Microsoft, please add this feature or respond to us why y'all aren't going to add it.

@bpasero added the Backlog milestone in 23 of October for this feature.

That's already some progress, there is a plan to do it, now we wait. :)

**Comment by DLiblik** [comment]
Highly-condensed version of wrapping tabs using @stevenlaidlaw approach... some space lost to the right due to editor commands, but seeing the tabs all at once is worth it 😃 

```css
.tabs-and-actions-container>.monaco-scrollable-element {
    height: auto !important;
}

.tabs-and-actions-container>.monaco-scrollable-element>.tabs-container {
    height: auto !important;
    flex-wrap: wrap;
}

.tabs-and-actions-container>.monaco-scrollable-element>.tabs-container .label-description
 {
    margin-left: 0;
}

.monaco-workbench .part.editor>.content .editor-group-container>.title .tabs-container>.tab .monaco-icon-label:before {
    height: 35px;
}

.monaco-icon-label>.monaco-icon-label-container {
    flex-direction: column-reverse;
    display: flex;
}

.monaco-icon-label:before {
    background-size: 20px;
    width: 20px;
}

.monaco-workbench .part.editor>.content .editor-group-container>.title .tabs-container>.tab {
    width: 80px;
    min-width: fit-content;
    min-width: -moz-fit-content;
    flex-shrink: 0;
}

.monaco-workbench .part.editor>.content .editor-group-container>.title .tabs-container>.tab>.tab-close {
    width: 15px;
}

.monaco-workbench .part.editor>.content .editor-group-container>.title .tabs-container>.tab>.tab-close .action-label {
    margin-right: 0;
}
```

![image](https://user-images.githubusercontent.com/4047997/72033126-6110e300-325f-11ea-99cd-0cd27fdd6d3a.png)

**Comment by mmzoo** [comment]
For me it would already suffice if the tab bar area would be wider than 50% of the window, i.e. stretch all the way to the right (like in TextMate).

![Screen Shot 2020-02-07 at 09 53 10](https://user-images.githubusercontent.com/1029840/74016971-9f4f0e80-4993-11ea-964d-4686a6f23c19.jpg)

I could not find an issue for this feature. But this issue here comes closest.

**Comment by fraigo** [comment]
This is what I've done after a couple of tweaks. 
I noticed when changing the theme, you need to "_Reload Custom CSS and JS_" again.

```css
.monaco-workbench .part.editor>.content .editor-group-container>.title.tabs>.tabs-and-actions-container>.monaco-scrollable-element {
    height: initial !important;
}

.monaco-workbench .part.editor>.content .editor-group-container>.title .tabs-container{
    height: initial !important;
    flex-wrap: wrap;
    justify-content: space-between;
}

.monaco-workbench .part.editor>.content .editor-group-container>.title .tabs-container>.tab{
    height: 30px;
    flex-grow: 1;
    border-bottom: 1px solid rgba(128,128,128,0.5);
}
```

![image](https://user-images.githubusercontent.com/8151943/76644826-002eb100-6515-11ea-8da0-e17efa75b4e6.png)

**Comment by J3m5** [comment]
Still not planned for [April](https://github.com/microsoft/vscode/issues/95168), seriously, come on...

**Comment by arnotes** [comment]
yes please :)

**Comment by dgwyer** [comment]
Surprised this wasn't available already as VS Code is so flexible. Fingers crossed it can be added soon. I always have too many tabs to fit on one line and scrolling to find tabs is a real pain!

**Comment by lucaszanella** [comment]
It's almost as if vscode developers never used vscode for actual development

-------- Mensagem Original --------
Ativo 16 de abr de 2020 15:49, David Gwyer escreveu:

**Comment by jarileon** [comment]
Started using VS code because a React course suggested it. One of the first things I tried to enable was tabs on multiple rows - but the option is missing, which is a real bummer. The actual Visual Studio has it and I use it all the time. VS code needs this ASAP!

**Comment by dgwyer** [comment]
Interesting that Visual Studio already has this.

> Started using VS code because a React course suggested it. One of the first things I tried to enable was tabs on multiple rows - but the option is missing, which is a real bummer. The actual Visual Studio has it and I use it all the time. VS code needs this ASAP!

**Comment by lazzyn** [comment]
Oh common Microsoft, do it already. This has been in Visual Studio for ages.

**Comment by im-richard** [comment]
I'm going to have to start adding an additional 500 calories to my diet just because my finger is having to fight with the constant scrolling so much.

It may seem like a small feature to them, but it has actually caused me to contemplate moving editors. The only reason I haven't is because I have VSCode setup exactly how I want it, and screwing with a new one right now is just going to lag my workflow / performance for a bit.

I can't imagine such a feature would take much work, and it would make a lot of people happy.

**Comment by J3m5** [comment]
Get ready for the next iteration plan! 💪 ✊

**Comment by J3m5** [comment]
It's [here](https://github.com/microsoft/vscode/issues/97432), pinned tabs are planned but still no multi line tabs... 😢

**Comment by im-richard** [comment]
Not shocked anymore. I've come to the conclusion that these guys don't use their own software. There's no way that a developer who uses this on an every-day basis AND has pull with deciding what VSCode has; could not be irritated at constantly scrolling.

**Comment by bedge** [comment]
Re: 
> Highly-condensed version of wrapping tabs using @stevenlaidlaw approach... some space lost to the right due to editor commands, but seeing the tabs all at once is worth it 😃
> 
> ```css
> .tabs-and-actions-container>.monaco-scrollable-element {
>     height: auto !important;
> }
> 
> .tabs-and-actions-container>.monaco-scrollable-element>.tabs-container {
>     height: auto !important;
>     flex-wrap: wrap;
> }
> 
> .tabs-and-actions-container>.monaco-scrollable-element>.tabs-container .label-description
>  {
>     margin-left: 0;
> }
> 
> .monaco-workbench .part.editor>.content .editor-group-container>.title .tabs-container>.tab .monaco-icon-label:before {
>     height: 35px;
> }
> 
> .monaco-icon-label>.monaco-icon-label-container {
>     flex-direction: column-reverse;
>     display: flex;
> }
> 
> .monaco-icon-label:before {
>     background-size: 20px;
>     width: 20px;
> }
> 
> .monaco-workbench .part.editor>.content .editor-group-container>.title .tabs-container>.tab {
>     width: 80px;
>     min-width: fit-content;
>     min-width: -moz-fit-content;
>     flex-shrink: 0;
> }
> 
> .monaco-workbench .part.editor>.content .editor-group-container>.title .tabs-container>.tab>.tab-close {
>     width: 15px;
> }
> 
> .monaco-workbench .part.editor>.content .editor-group-container>.title .tabs-container>.tab>.tab-close .action-label {
>     margin-right: 0;
> }
> ```
> 
> ![image](https://user-images.githubusercontent.com/4047997/72033126-6110e300-325f-11ea-99cd-0cd27fdd6d3a.png)

@DLiblik 
I like the idea, but there's too much padding around each filename, any suggestion for compressing a bit?
Also, the wrapping is non-optimal:

![Screen Shot 2020-05-15 at 7 45 41 AM](https://user-images.githubusercontent.com/499317/82063300-280ad400-9680-11ea-8838-47c6df29bdcf.png)

I know it's the toolbar that's causing the wrapping issue, and I've been using this as a console cmd to remove it:
```
[...document.querySelectorAll('.editor-actions')].map(a => a.style.display = 'none')
```
But if we're wrapping, it would be preferable to keep the toolbar and extend to the full width below it.

**Comment by DLiblik** [comment]
> I like the idea, but there's too much padding around each filename, any suggestion for compressing a bit?

Unfortunately @bedge the DOM structure/classes for the info in the tabs has changed since I did that - the folder name showed at the top when I did that up.

I was going to try and keep up with VS Code changes, but then the customizations had VSCode warning about non-standard code (at least at last check) which caused other issues, so I just gave up and went back to the out-of-box single row line-up.  When I'm up to my ankles upside-down in concurrent edits  then I just use a different editor for now.  

And no complaints, VS Code is, after all, free and it's a great tool, but it's definitely sucking a little wind on this item right now.  I'd jump in and PR it myself but I'm heavily into some other community undertakings atm.

**Comment by lazzyn** [comment]
> It's [here](https://github.com/microsoft/vscode/issues/97432), pinned tabs are planned but still no multi line tabs... 😢

This is just sad.

**Comment by jay4110** [comment]
C'mon Microsoft you can do it. JUST DO IT

**Comment by jmoois** [comment]
<Please read this with a light, semi-joking tone, not a mad tone>

Hey Microsoft, could you please give us some sort of feedback. If you're not planning on doing multirow tabs can you tell us, or maybe "some day", or "it's on the backlog somewhere", or "we're looking into what it might take"... or....   some sort of information so at least we know you are listening and paying attention. Developers hate it when they feel like their tool makers are paying no attention to them. Let us know you're there =)

**Comment by PachowStudios** [comment]
Dear corporate overlords, please hear our cry.

**Comment by scorsy63** [comment]
+1 please add this feature

**Comment by lazzyn** [comment]
Microsoft, please. I promise to buy a Zune if you ever release it again. Just add this feature, please!!

**Comment by J3m5** [comment]
Please, don't forget to upvote the first post of this issue with a 👍.
 Thank you.

**Comment by lazzyn** [comment]
> Please, don't forget to upvote the first post of this issue with a 👍.
> Thank you.

Yes, yes and yes!!

**Comment by yulei900609** [comment]
yes, please.

**Comment by zonesql** [comment]
Please add this feature, or provide an update on this product request.

**Comment by im-richard** [comment]
Captains log... #295. Still no multi-row support. When will the madness ever end.

**Comment by scorsy63** [comment]
I tried the solution posted by @stevenlaidlaw and it works perfect.  Just one small issue, after installing the required extension I get a "[Unsupported]" message in the title bar.  Other than that, it works just like I wanted it to.

**Comment by ftuyama** [comment]
The VScode upgrade broke the feature https://stackoverflow.com/questions/42462777/multirow-tabs-for-vscode
I'm so unhappy right now... 😢

**Comment by msesen** [comment]
This is really bad, especially when doing Angular development. This is counter-productive, does anyone has any other IDE recommendations please?

**Comment by rajnisheu** [comment]
this is bad when doing _any_ development. Go for jetbrains. its worth it. with such a basic feature missing, no plans to implement, vscode is more of a toy.

**Comment by boumabcd** [comment]
hello guys, same as everyone the lack of feedback is quite annoying...

i did a little powershell script for windows users to automate the css change (thus you dont need the VSCode Custom CSS extension and all the shenanigans). It can easilly be adapted for linux/mac users using bash instead, just be carefull with the encoding when you write to the css file (on windows it seems to be UCS-2 LE BOM according to notepad++)
the file which will be updated is VSCODEINSTALLDIR\resources\app\out\vs\workbench\workbench.desktop.main.css

how to use:
-close vscode
-adapt the following variables: $vscodeInstallPath and $customCssStr
-if vscode was installed globally, run the script as admin
-restart vscode, you're good to go

the script:
```powershell
# change this by your installation path of vscode
$vscodeInstallPath = "C:\Program Files\Microsoft VS Code"

# you can change the css by somethig else but if you want to change the css code, you'll have to manually remove the previous css code
$customCssStr = @"
 .tabs-and-actions-container > .monaco-scrollable-element {
  height: auto !important;
}

.tabs-and-actions-container > .monaco-scrollable-element > .tabs-container {
  height: auto !important;
  flex-wrap: wrap;
}
"@



$cssComment = "/* MANUAL EDIT: custom css for multi line tabs */"
$customCssStr = "`n" + $cssComment + "`n" + $customCssStr

$mainCssPath = "resources\app\out\vs\workbench\workbench.desktop.main.css"
$fullmainCssPath = join-path $vscodeInstallPath $mainCssPath

$cssFile = gi $fullmainCssPath -ErrorAction SilentlyContinue
if ($cssFile) {
	if ($(Select-String -InputObject $(gi $fullmainCssPath) -SimpleMatch $cssComment) -eq $null) {
		write-host "adding custom css and making a backup of the file $fullmainCssPath"
		Copy-Item -LiteralPath $fullmainCssPath -Destination $($fullmainCssPath + ".old")
		$customCssStr | Out-File -LiteralPath $fullmainCssPath -Append
	} else {
		write-host "custom css already in the file, please remove it manually"
	}
} else {
	write-host "file $fullmainCssPath not found"
}
# to revert the modification simply use
# Copy-Item -LiteralPath $($fullmainCssPath + ".old") -Destination $fullmainCssPath
```

**Comment by ChristianWilkie** [comment]
@microsoft any update on this?

**Comment by rocifra** [comment]
@microsoft I am also interested in an update on this. I've been waiting on it for over an year now since I started working on a relatively large project and each day I'm working with many opened tabs/files (json, interfaces, classes) which are mostly interlinked and prefer to always keep them opened to juggle easier between them.

**Comment by vtcifer** [comment]
> I've implemented a stop-gap solution using [VSCode Custom CSS](https://github.com/be5invis/vscode-custom-css). Just create your own custom CSS file in your system and fill it with the following:
> 
> ```css
> .tabs-and-actions-container > .monaco-scrollable-element {
>   height: auto !important;
> }
> 
> .tabs-and-actions-container > .monaco-scrollable-element > .tabs-container {
>   height: auto !important;
>   flex-wrap: wrap;
> }
> ```
> 
> Then just point the plugin at your custom file in your settings.json like so:
> 
> ```json
>     "vscode_custom_css.imports": [
>         "file:///home/username/path/to/dir/custom.css"
>     ],
>     "vscode_custom_css.policy": true,
> ```
> 
> Then hit `ctrl-shift-p` and run "Reload custom CSS and JS"
> 
> This has solved the problem for me until they implement an in-app solution.

I've used this work around, but it seems to break the horizontal scroll bar when the tabs wrap...

**Comment by aureliusm** [comment]
@microsoft That would be a really really helpful feature! One of the bigger things I miss in VS code.

**Comment by ErnirSH** [comment]
Anyone know if they are working on this ?

**Comment by ciltocruz** [comment]
> > I've implemented a stop-gap solution using [VSCode Custom CSS](https://github.com/be5invis/vscode-custom-css). Just create your own custom CSS file in your system and fill it with the following:
> > ```css
> > .tabs-and-actions-container > .monaco-scrollable-element {
> >   height: auto !important;
> > }
> > 
> > .tabs-and-actions-container > .monaco-scrollable-element > .tabs-container {
> >   height: auto !important;
> >   flex-wrap: wrap;
> > }
> > ```
> > 
> > 
> > Then just point the plugin at your custom file in your settings.json like so:
> > ```json
> >     "vscode_custom_css.imports": [
> >         "file:///home/username/path/to/dir/custom.css"
> >     ],
> >     "vscode_custom_css.policy": true,
> > ```
> > 
> > 
> > Then hit `ctrl-shift-p` and run "Reload custom CSS and JS"
> > This has solved the problem for me until they implement an in-app solution.
> 
> I've used this work around, but it seems to break the horizontal scroll bar when the tabs wrap...

This does not work in vscode installed by Snap. Because the system is mounted in read-only mode.
Any alternative solution?

**Comment by im-richard** [comment]
I'm using the css edits that were posted here. It does help the issue and at least I have multi-row tabs now instead of one scrolling bar, but vscode wants to complain for some reason that the install is corrupt.

No idea why, but it goes away when I remove the configuration for custom css. So for now I've just set the error to ignore.

Honestly pretty sad that we have to ghetto-rig something so... basic. Something I'd imagine that would take almost no time at all for the VSCode team to include, but yet push features that I've never seen a single recommendation for.

That one issue was obstructing my workflow quite a great deal; so I can only imagine the frustration others are experiencing.

**Comment by danfmaia** [comment]
> PLEASE, THIS FEATURE IS EXTREMELY USEFUL
> 
> Look at my firefox: https://imgur.com/a/CKOGGMq
> 
> I can't live without miltiline tabs
> 
> It's VERY annoying to search for the opened tabs by scrolling. I'm almost giving up on vscode because of this.

You could also pin the most used tabs to save space.

**Comment by scorsy63** [comment]
For those that use the [VSCode Custom CSS](https://github.com/be5invis/vscode-custom-css) solution posted here, you can remove the "unsupported" message that VS Code complains about with this exension [Fix VSCode Checksums](https://marketplace.visualstudio.com/items?itemName=lehni.vscode-fix-checksums)

**Comment by mike-boost** [comment]
Please add multi-row tabs feature!

**Comment by im-richard** [comment]
With the Custom CSS solution, I've also noticed that it makes the horizontal scrollbar disappear as well. Not sure if anyone else is experiencing this, but as soon as I load it, the bar goes away.

**Comment by lokize** [comment]
+1 for this

**Comment by braddialpad** [comment]
As someone who uses a vertical monitor fairly regularly, this is very important to have.

**Comment by yulei900609** [comment]
Please add multi-row tabs feature!

**Comment by apereyra** [comment]
+1 for this

**Comment by pdxwolfy** [comment]
Please. +1

**Comment by SneakyFish5** [comment]
Started working on this, huge props to @stevenlaidlaw cause I suck at CSS 😄. Commit is [here](https://github.com/SneakyFish5/vscode/commit/1e2991a7061f0c6d8137f61049a5f51f9bd0a24f), it would be amazing if someone was willing to clone my fork and test it out, to make sure there aren't any issues. If I'm missing something CSS wise, please let me know. I didn't include any of the other proposed style changes, since I assume those should be in a different issue entirely.

![QX0AkGgNj9](https://user-images.githubusercontent.com/32284796/92549380-b7835680-f21e-11ea-952a-8f966b61ef44.gif)


It currently requires a window reload before taking effect, however I can try to figure out how to have it automatically wrap when changing the setting if it's requested.

**Comment by vannaka** [comment]
@SneakyFish5 A big problem with the custom CSS hacks above is when trying to rearrange the tabs by dragging and dropping the drop zone would not be adjusted to cover the new rows of tabs so you would be able to start dragging any tab but could only drop it on the first row. Do your changes account for this by expanding the drop zone to cover all tabs no matter how many rows exist?

**Comment by SneakyFish5** [comment]
@vannaka I was unaware of that problem, but after testing it out yes, that is an issue. Like I said, I'm really awful at CSS (and I'm assuming its a CSS issue?). I messed around with it and couldn't figure it out. The CSS file for tabs is [here](https://github.com/microsoft/vscode/blob/master/src/vs/workbench/browser/parts/editor/media/tabstitlecontrol.css). If anyone comes up with a solution for it, I can work on applying it on the JS side of things.

**Comment by vannaka** [comment]
@SneakyFish5 I found the html/css for the drop zone. I used the dev tools available via `Help -> Toggle Developer Tools` to find the div element for the drop zone. The html/css for the drop zone is generated dynamically at runtime and is (probably) triggered by drag/hover events and as such you'll need to find the typescript that handles these events and change the css they generate.

![drop_zone_css](https://user-images.githubusercontent.com/5122956/92638783-716ed700-f2a0-11ea-8e31-6f490816d982.PNG)

As you can see in the screenshot, the `top` and `height` attributes are incorrect. I can change them manually and watch the drop zone shift down as desired. You'll need to find the code that sets `style="top: Npx; height: calc(100% - Npx);"` to set the height offset appropriately on the `div` html element (red circle in screenshot). I'd suggest grepping through the code for the beginning of the above `style=...` string to get you started although it's likely the `35px` is not hardcoded and thus you can't just search for the whole string..

**Comment by fraigo** [comment]
@SneakyFish5 I've made a change to set the context of multi-line tabs to a higher level container (element `.tabs-and-actions-container`), to fix cases when the `.monaco-scrollable-element` inside is setting a fixed container height _when opening a new file_ (otherwise, the css fix has no effect into the inner container). Also adding a light bottom border for tabs.

https://github.com/SneakyFish5/vscode/commit/a38aaa56611bb652d950a5768cdabcb671ecb88a

Also, as suggested @vannaka, I added a fix to the DropTarget element to dynamically calculate  the height of the tab section (current code uses a fixed height of 35)

https://github.com/SneakyFish5/vscode/commit/f0695d51cf22af967946674acbd1212d082a5f91

**Comment by SneakyFish5** [comment]
That was so helpful, thank you both so much, saved me a lot of time and almost definitely a CSS-induced headache haha. I implemented a listener as well to watch for settings changes, so a window reload is no longer required and it'll smoothly transition into using multiline tabs.

![M5b7NlCmUZ](https://user-images.githubusercontent.com/32284796/92657121-dbe04100-f2b9-11ea-91ad-affa340a50d8.gif)

Everything look good?

**Comment by Manachi** [comment]
@SneakyFish5 You, my friend, are an absolute champion - thank you for your incredible work on this!

**Comment by SneakyFish5** [comment]
@Manachi thank you thank you, I just got the ball rolling though. Couldn't have done it without everyone here especially @fraigo, @vannaka, and @stevenlaidlaw. It was a group effort for sure. I'll open up the PR by end of day tomorrow, just want to do last minute testing and such.

**Comment by lazzyn** [comment]
> That was so helpful, thank you both so much, saved me a lot of time and almost definitely a CSS-induced headache haha. I implemented a listener as well to watch for settings changes, so a window reload is no longer required and it'll smoothly transition into using multiline tabs.
> 
> ![M5b7NlCmUZ](https://user-images.githubusercontent.com/32284796/92657121-dbe04100-f2b9-11ea-91ad-affa340a50d8.gif)
> 
> Everything look good?

This looks awesome, thanks mate 👍

**Comment by SneakyFish5** [comment]
PR opened 🎉

**Comment by pdxwolfy** [comment]
Thank you so much for doing this!

**Comment by swinder0161** [comment]
Is this coming into insider or release channel anytime sooner?

**Comment by Jeyloh** [comment]
* holding breath

**Comment by jmoois** [comment]
Any news?  Does Microsoft plan to accept SneakyFish5's PR? If not, could you tell us why not? If so, do you have an even rough timetable you could share?

Thanks.

**Comment by vannaka** [comment]
@jmoois and everyone else, if you want to follow progress please head over to the  [PR page](https://github.com/microsoft/vscode/pull/106448) and read through the thread there instead of spamming this thread with never ending "is it done yet?" comments. I hate getting notifications for this issue only to see it just some donut who's too lazy to do a little leg work for themselves.

**Comment by fraigo** [comment]
The best way to contribute and keep this going on right now is to clone/fork the repo and play with this version, spotting any issues and, if possible, posting code fixes. The same if you can't wait for a release to start using it. 
A quick start script:
```
git clone https://github.com/SneakyFish5/vscode.git
git checkout wrap-tabs
cd vscode
yarn install
./scripts/code.sh
```

**Comment by SneakyFish5** [comment]
Hey everyone following this issue, very sorry for a lack of progress. I've been swamped recently with work, but rest assured I will get around to it (hopefully this weekend again). This is a feature I'd really like to have myself 😄. This is 100% on me and @bpasero has been extremely helpful every step of the way. And like @fraigo said, the best way to help is to contribute code fixes/spot any issues or be a little bit more patient.

**Comment by jmoois** [comment]
I didn't mean to offend anyone. I'm sorry. I was just worried this was turning into a still-born feature (Like so many features in other products I've cared about over they years). While I had looked at the actual PR page, the last time I had read it there wasn't much activity so I thought this thread was much more metalevel and therefore a better place to ask more metalevel questions (as opposed to to specifics of actual code in the PR).
But again, I'm sorry.

**Comment by jmoois** [comment]
@fraigo 

I tried installing but when I got to the yarn install step, I get the following:

yarn install
yarn install v1.22.5
$ node build/npm/preinstall.js
[1/4] Resolving packages...
[2/4] Fetching packages...
info fsevents@2.1.3: The platform "win32" is incompatible with this module.
info "fsevents@2.1.3" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@2.1.1: The platform "win32" is incompatible with this module.
info "fsevents@2.1.1" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@1.2.7: The platform "win32" is incompatible with this module.
info "fsevents@1.2.7" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
warning " > eslint-plugin-mocha@8.0.0" has incorrect peer dependency "eslint@>=7.0.0".
warning " > gulp-tsb@4.0.5" has incorrect peer dependency "typescript@^3.0.0".
[4/4] Building fresh packages...
[1/19] ⠐ keytar
[2/19] ⠐ native-is-elevated
[3/19] ⠐ native-keymap
[4/19] ⠐ native-watchdog
error C:\programs\vscodeMultiTabs\vscode\node_modules\native-keymap: Command failed.
Exit code: 1
Command: node-gyp rebuild
Arguments:
Directory: C:\programs\vscodeMultiTabs\vscode\node_modules\native-keymap
Output:
C:\programs\vscodeMultiTabs\vscode\node_modules\native-keymap>if not defined npm_config_node_gyp (node "C:\Program Files\nodejs\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "" rebuild )
gyp info it worked if it ends with ok
gyp info using node-gyp@3.8.0
gyp info using node@12.0.0 | win32 | x64
gyp ERR! configure error
gyp ERR! stack Error: Command failed: C:\Python38\python.EXE -c import sys; print "%s.%s.%s" % sys.version_info[:3];
gyp ERR! stack   File "<string>", line 1
gyp ERR! stack     import sys; print "%s.%s.%s" % sys.version_info[:3];
gyp ERR! stack                       ^
gyp ERR! stack SyntaxError: invalid syntax
gyp ERR! stack
gyp ERR! stack     at ChildProcess.exithandler (child_process.js:304:12)
gyp ERR! stack     at ChildProcess.emit (events.js:196:13)
gyp ERR! stack     at maybeClose (internal/child_process.js:1000:16)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:267:5)
gyp ERR! System Windows_NT 10.0.18363
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\programs\vscodeMultiTabs\vscode\node_modules\native-keymap

**Comment by fraigo** [comment]
@jmois please see this document about prerequisites for the dev build:
https://github.com/Microsoft/vscode/wiki/How-to-Contribute#prerequisites

You could be missing something, especially the compiler tools for win32 (used by node-gyp), for example:
`npm install --global windows-build-tools --vs2015`

If it still does not work please refer to the node-gyp installation prerequisites
https://github.com/nodejs/node-gyp#on-windows

**Comment by jgbolling** [comment]
Just put your cursor over the tab bar and use the mouse wheel to scroll the tab bar horizontally to see the tabs that aren't otherwise visible.

**Comment by AlonMiz** [comment]
+1

**Comment by jmoois** [comment]
ok, with the windows-build-tools it now builds successfully. I am seeing an
interesting issue.  If:
1)  I don't yet have enough tabs to get multiple rows, and
2) I expand the window wider horizontally than the tab bar,
then
result:  the content/document area area goes blank.
I can switch tabs and the content/document area is still blank.
The only thing that gets the content visible again is decrease the width of
the window to less than the width of the tab bar.

Also, if you change the width of the tab bar (by opening a new one or
closing one) you get the same situation with the new width.

Jeff

**Comment by SneakyFish5** [comment]
Are you able to reproduce this with the latest changes? Tested it and I wasn't able to when resizing the window, however I might be doing it wrong. If you are able to repro, do you mind providing a recording of what you mean?

**Comment by jerviscui** [comment]
I want this function too.

Will the official merge this?

**Comment by swinder0161** [comment]
Finally, I moved to Ultraedit(Paid Software) to get this feature. It was a show stopper for me along with a few more features which are poorly written in VSCode.

**Comment by aborisovich** [comment]
Guys... Please?

**Comment by bpasero** [comment]
Thanks for the continued interest in this feature and sorry for not having posted back here, I wanted to give a small update of where we are with this request.

### Current Status

In https://github.com/microsoft/vscode/pull/106448, @SneakyFish5 and me have been working on getting an initial version working. This is a lot more than just a CSS change because the editor container below the tabs needs to be sized properly and with tabs wrapping, the height is now dynamic when it was static before. I would say, the PR is now in a state where the basic functionality of tabs wrapping is working OK.

However, there are some UX concerns after presenting this in the team that I would consider required before merging this in. Specifically:
* we feel that the editor actions in the top right corner should move to the bottom (into the last row when tabs wrap) and free up the space above so that tab-rows can fill the entire width. with many tab rows, we otherwise have lots of wasted space for no good reason and this get's worse if you have different editors use different number of actions because this causes tabs to move around (_[1]_)
* we currently decided to let a tab fill the entire row, even if that makes the tab very large. maybe in this case the tab label should be centered and not left aligned, but specifically when `workbench.editor.tabSizing: shrink` is configured, ideally all tabs get the same width so that we can get a look & feel similar to what Firefox seems to be doing (_[2]_)
* we currently have no limit as to how many tab rows you can get, but we feel that you should probably never see more than 3-4 rows of tabs, and then we fallback to showing a single row if the number of tabs gets too large

On top of that:
* I find the idea of a theme-able border between tab rows interesting (https://github.com/microsoft/vscode/issues/70413#issuecomment-598836942) and worthwhile to explore
* I noticed many comments with CSS tweaks and would be open to take some of them into the final solution when tabs wrap, e.g. to produce smaller tabs or even center the label if a tab grows large in an otherwise empty row

_[1]_
![image](https://user-images.githubusercontent.com/900690/101748341-b509f480-3acc-11eb-91e1-e894b77f0504.png)

_[2]_
![image](https://user-images.githubusercontent.com/900690/101748357-bcc99900-3acc-11eb-9ee3-090ae9f99a6e.png)

### Alternatives

I would also like to understand why "Open Editors" view is not sufficient for this task and/or wether having vertical tabs as an option would be a better solution overall. Visual Studio recently added support for vertical tabs to appear either left or right:

![image](https://user-images.githubusercontent.com/900690/101748685-26e23e00-3acd-11eb-8b60-108a5155986c.png)

I try to understand if we could implement the one or the other and if we really need both.

### How we use tabs/editors

Some users here have asked how we in the team survive without tabs wrapping. I think there is no single answer here, but I wanted to respond to that with some advice what we are doing:
* a lot of us actually have decided to disable tabs (`workbench.editor.showTabs: false`) not because of missing wrapping but because we use other ways of navigating editors (see below) and many of us find tabs to be adding UI clutter that is overwhelming
* even when tabs are not disabled, keybindings and quick pick offer nice alternatives to navigate and find tabs without having to use the mouse and click on a tab (_[3]_). 
* the "Open Editors" allows to navigate the list of editors even when tabs are disabled or scrolled out of view

_[3]_
![recording](https://user-images.githubusercontent.com/900690/101749586-357d2500-3ace-11eb-988e-fe96550fc3da.gif)

**Comment by yulei900609** [comment]
So, no tab wrap in the future?

**Comment by bpasero** [comment]
> So, no tab wrap in the future?

Maybe my comment was not clear: the PR is not done, anyone willing to help out should contribute to it (please coordinate with SneakyFish5).

Besides, if everyone thinks, vertical tabs are inferior over wrapping tabs, I think we should have a discussion which solution we are going after for. I am open to hear feedback on that front here in this issue.

**Comment by im-richard** [comment]
I personally would be a lot more happy with wrapping tabs over vertical due to the addons I use and where they are. I tried using the vertical list of open files and I always go back to the top just because it's way easier. The only downside is the scrolling. If tabs wrapped, then I could control how many rows I'm able to deal with and it just makes the workflow a lot more smooth.

Others may have their own preference, but for me personally; I'll always vote for wrapping tabs over the alternatives.

**Comment by GonziHere** [comment]
@bpasero 
I use 4K 40" with 3 code panels (columns) at once. If I would use vertical tabs, I would either have several, mostly empty columns so I would lose real estate for my code, or I would have one column (visual studio implementation) with several groups, which I have trouble navigating (which group is which column of code?). It still takes precious real estate.

Horizontal tabs are used by default everywhere because they are optimal both visually, and spatially, but they need to add rows for more opened documents. If you have 20 tabs, you are most likely doing something wrong, but If you have like 4 documents with longer filenames (or you need to see part of the path), you would likely prefer to lose just one line of code real estate to the ability to switch instantly, to see if you have or have not this file open, etc.

To each his own, but that's why I don't use any vertical tabs anywhere. Horizontal ones take almost no space (and I can spare the row, or rows), but vertical ones cut the viewport width, and I cannot have that. This is why I also routinely show/hide the left sidebar with ctrl-b, or use (inferior) "floating sidebar" option in visual studio.

**Comment by KrisPetkov** [comment]
I'm also voting for wrapping horizontal tabs. I don't think that vertical tabs are more useful. I also think that 90% of the people are familiar and used to horizontal tabs. The vertical ones seems as a corner case and rarely used feature to me.

**Comment by dgwyer** [comment]
Wrapping horizontal tabs all the way for me! I doubt I'd use vertical tabs.

**Comment by csears123** [comment]
The Open Editors panel (vertical tabs) are not a desired coding UI workflow for me.  I have several vertical panels on the left side that currently manage other important coding tools (Workspace, Outline, SSH File Systems, etc.) and trying to fit another panel for vertical tabs is going to make the space on the left side too small to work with, constantly needing to collapse and expand panels, etc.
I also tend to code towards the top center of my screen, which is where my open tabs are as well, so the location convenience of tabs at the top is my preference to switch between open files/editors. Needing to access the Open Editors panel to the left is not as fast and convenient.
I understand there are quick keybindings that can help switch, but as a keyboard and mouse user I find myself needing to use the mouse to navigate, select, and/or position the cursor once I enter the desired file/editor.  So relying on just the keyboard bindings is not (for me) sufficient anyways.
Allowing the tabs at the top to wrap would be the desired workflow for me, for best efficiency.

**Comment by illz** [comment]
The vertical tabs are not useful to me; too many years with normal wrapping
horizontal tabs, which get my vote.

**Comment by im-richard** [comment]
It's also as @GonziHere stated; vertical tabs also cost precious real-estate. 9 times out of 10; I close the left-side interface because I like my workspace left available for my code. The only time I ever open anything in that bar is when I'm utilizing the search feature to find results in multiple files.

Thanks to other apps; horizontal tabs are familiar; and take up little space.

Of course, it would also be a nice addition if horizontal tabs had title settings such as:

- Display full title of opened tab
- Truncate titles if more than X open ( My Pro... would replace a tab called My Project if amount of tabs exceeds workspace width )

But again, that's just a cherry on the top; just having Horizontal tabs alone would be a huge leap forward.

**Comment by bpasero** [comment]
Thanks for all the feedback, that is very valuable. Thinking more about it I actually feel having vertical tabs as an option is not making so much sense given we already have the "Open Editors" view which is pretty much that, but with the benefit of having it inside a view that can be collapsed, moved and even hidden.

Let's push forward on the PR then and get it into a state that it can be merged.

**Comment by vtcifer** [comment]
> * we currently decided to let a tab fill the entire row, even if that makes the tab very large. maybe in this case the tab label should be centered and not left aligned, but specifically when `workbench.editor.tabSizing: shrink` is configured, ideally all tabs get the same width so that we can get a look & feel similar to what Firefox seems to be doing (_[2]_)

I am *really* not a fan of either of these.  A tab's size / alignment really shouldn't depend on how many other tabs are open, when tabs are allowed to wrap.

**Comment by bpasero** [comment]
@vtcifer if tabs are not filling the rows, you will end up with a weird pattern where some tabs leave gaps at the end and the overall look and feel is not very pleasant. If tabs fill these gaps by getting larger, it looks more homogeneous. Happy for ideas how to solve this though if you have any.

**Comment by illz** [comment]
Options!
expand tabs to fill row
expand tabs to fit file name
expand tabs to fit file name + path
truncate tab file names to x characters
truncate tab file names to x percent of width / set max row tab count

Choose a base set of release options, over time expand to cover more
scenarios. Build truncation, width, and row tab count as separate variables
to enable a wide range of usage.
We're talking about variations on displaying an click-actionable list.
There's a pretty well explored number of variations on list displays,
whether horizontally stacked, multi-row, variable width, etc. Let's just
get a basic set of knobs to adjust and go from there. It has literally been
years since this conversation started and yet we have nothing; seems a bit
silly when the solution can start simply and expand to meet almost all
desires.

**Comment by vtcifer** [comment]
>  ...the overall look and feel is not very pleasant...

This is an opinion, and much like unmentional parts of the body, everyone has one, and they all stink.  

Personally, I see nothing wrong with this:

![image](https://user-images.githubusercontent.com/12104547/101996227-6f9d2100-3c9e-11eb-9fdb-0fee001b031f.png)

Honestly a user configurable setting, one option being "Don't do anything special" would be the best way imo.  _Similar_ to most text processing applications giving the options for alignment "Left Align, Center, Right Align, Justify".

**Comment by pdxwolfy** [comment]
I also see absolutely nothing wrong with that. I have Firefox doing the same multi-tab arrangement (with help from `userChrome.css`). I had Atom and Sublime Text configured the same way. +1 for multi-row tabs that are just wide enough, and the heck with the extra space on the right.

**Comment by cfxd** [comment]
Thank you to everyone who's been working hard on this and keeping this feature request moving in the right direction.

I eagerly look forward to this feature and being able to read my open file names again (soon) (hopefully)! 🚀

**Comment by bpasero** [comment]
https://github.com/microsoft/vscode/pull/106448 has been merged to `master` and this feature will be available starting with tomorrows insider build through a new setting `workbench.editor.wrapTabs`. Please give it a try then and report back how it goes.

A few more changes went in over the past weeks:
* editor actions now appear bottom right to avoid the large gap we had before
* the limit of max 3 rows was removed (tabs will wrap until the editor reaches a minimal height)
* `tab.border` theme color is enabled at the bottom of tabs when tabs wrap as separator

![image](https://user-images.githubusercontent.com/900690/103521584-4cb6f480-4e79-11eb-921a-8226cdc539d4.png)

**Comment by vannaka** [comment]
@SneakyFish5 you are a hero!

**Comment by christiancolemandeloitte** [comment]
Ha! Only just now googled this as a pain point for myself and I'm lucky enough to see that it's going live very soon. Awesome!

**Comment by swinder0161** [comment]
@bpasero @SneakyFish5 Thanks for your efforts to have this feature merged. I am already using and loving it.
Just wondering if there is option to change the tab height? it looks like tab height is too much on mac version.

Also the tabs are not equal size, it makes the file selection a bit confusing. And the last line of tabs is in justified text alignment mode, so a single tab or 2 tabs cover the whole line.

Is there an option to have left aligned tabs and leave empty space on the right?

**Comment by bersbersbers** [comment]
I have the following issue today on Version: 1.53.0-insider (user setup), Commit: a48ef56fbf7767f55a193d7d5687addd1fc9239c. When activating an editor group, additional icons may appear, changing the line wrapping of editor tabs and making the text jump.

![0qNq6eyLmr](https://user-images.githubusercontent.com/12128514/103617511-a7f6ee80-4f2e-11eb-8904-69e6eab6104c.gif)

Should I file this separately or this a good place?

**Comment by bpasero** [comment]
In today's insider build (https://code.visualstudio.com/insiders/) you can enable wrapping tabs via `workbench.editor.wrapTabs: true`. I suggest that if you encounter any issues, you report them as individual issues (such as @swinder0161 and @bersbersbers) and keep one issue to one topic (not multiple things). That makes it easier to discuss things individually. 

One setting that impacts the sizing of tabs is: `workbench.editor.tabSizing`. Set it to `shrink` to stop tabs from growing large when wrapping.

Thanks for testing this 👍

**Comment by im-richard** [comment]
Just because it has been a long time coming; trying it out now and it is a-maz-ing. 
Seriously; this just puts the cherry right on top of VSCode. 

It sounds like such a simple thing for people to hound about for so long; but it really makes all the difference in the world, and it will dramatically help flow and cut down on needing to scroll and find a tab when some of us can work on 20+ tabs at once in our projects.

Thanks for adding this.

**Comment by fraigo** [comment]
@SneakyFish5 @bpasero thank you for the continuous effort on this feature. Glad to see this option available today on Insiders! 
I was wondering if there is another configuration option to control the minimum shrink size until it wraps to a new line when using **Workbench > Editor : Tab Sizing** : `shrink`. 

![wrap_demo](https://user-images.githubusercontent.com/8151943/103677676-72a2ce80-4f37-11eb-8b7b-430b32fc27cf.gif)

**Comment by bpasero** [comment]
There is no setting currently to control the minimum width of a tab when setting the tab sizing to `shrink`. It is hardcoded to 80px.

**Comment by swinder0161** [comment]
I have actually added custom width in shrink mode as part of a new PR which
has title for custom height.

Please upvote the same.

**Comment by fraigo** [comment]
@swinder0161 please share the link to that PR to follow up, thanks!

**Comment by vannaka** [comment]
@fraigo  Here's the issue for configurable tab height: https://github.com/microsoft/vscode/issues/113832.

**Comment by s-kocher** [comment]
With many files opened and multi-line tabs enabled, the heigh can be used a lot, extreme example (but we never know, lot of legacy code exploration 😅 ) :
![image](https://user-images.githubusercontent.com/27820864/104081455-efcb8d80-522e-11eb-807b-937780319eab.png)

I remember the good old Tab Mix Plus extension on Firefox allowing to enable multi lines tab but with another option for the max displayed rows :
![image](https://user-images.githubusercontent.com/27820864/104081472-138ed380-522f-11eb-9edd-5b5d8dda24ff.png)
So when more than maximum allowed rows are needed, the rows are scrollable in vertical direction by hovering the tab area and using mouse wheel.

In my humble opinion (and as a tab hoarder dealing often with this problem),  it's the best balance to allow many tabs easy navigation and better than other solutions : 
- single line with unreadable tabs shrinked to fit all (chrome browser behavior for example)
- single line, tabs readable (fully displayed or reasonable shrink) but "infinite" scroll to navigate between the first and the last item
- multiple lines with all tabs displayed and readable with lot of vertical space used

**Comment by bpasero** [comment]
As I said before, if you have a suggestion or found a bug with wrapping tabs, best to open a new individual issue for each.

**Comment by MatthewS2077** [comment]
> 
> 
> Would love to have this as well. We use VS Code mostly for Angular work and we typically have multiple components (TS, HTML, CSS) and services (TS) open at a time, and we like to group relevant tabs together at the top.

This would be so useful! Ability to select multiple tabs and group them under single tab.  And ideally option to have VSCode automatically group files with the same name but different extensions (eg test.component.ts/test.component.html/test.component.scss) under the same tab.

**Comment by farteryhr** [comment]
still this: https://github.com/microsoft/vscode/issues/70413#issuecomment-519766211
here's my Just-Condensed™ mostly-works™ style based on previous comments in case you may like it: (sorry for mosaic cuz it's workplace code)
![image](https://user-images.githubusercontent.com/1175749/105319234-87ea4f00-5bff-11eb-9384-83130dfb8f64.png)
```css
.monaco-workbench .part.editor>.content .editor-group-container>.title .tabs-container>.tab .monaco-icon-label:before {
  height: 24px;
  background-size: 16px;
  width: 16px;
}

.monaco-workbench .part.editor>.content .editor-group-container>.title .tabs-container>.tab>.tab-actions .action-label {
  margin-right: 4px;
}

.monaco-workbench .part.editor>.content .editor-group-container>.title .tabs-container>.tab>.tab-close .action-label {
  margin-right: 0;
}

.monaco-workbench .part.editor>.content .editor-group-container>.title.tabs>.tabs-and-actions-container>.monaco-scrollable-element {
  height: auto !important;
  min-height: fit-content;
}

.monaco-workbench .part.editor>.content .editor-group-container>.title .tabs-container{
  height: auto !important;
  min-height: fit-content;
  flex-wrap: wrap;
}

.monaco-workbench .part.editor>.content .editor-group-container>.title .tabs-container>.tab{
  height:24px;
  min-width: fit-content;
  min-width: -moz-fit-content;
  flex-shrink: 0;
  border-bottom: 1px solid rgba(128,128,128,0.5);
  padding-left: 4px;
}

.monaco-workbench .part.editor>.content .editor-group-container>.title .tabs-container>.tab .tab-label {
  line-height: 24px;
}
.monaco-workbench .part.editor>.content .editor-group-container>.title .tabs-container>.tab>.tab-close {
  width: 15px;
}
```
there is problem with the height of editor instance though, but not big trouble so i didn't investigate more...

**Comment by SantiBailors** [comment]
> 
> 
> I've gave up with VSCODE - this feature should've been built in, even Eclipse has it. I'm using IntelliJ now as a lot of things come out of the box.

Are you sure Eclipse has it ? I ended up here while searching for info about how to have multiple tab rows in Eclipse, and the many mentions of this feature I found about Eclipse are all from people who say they would like Eclipse to have that feature. In my experience they are right that Eclipse doesn't have it, and I would really like to have it. How do you enable multiple tab rows in Eclipse ?

**Comment by sean-d** [comment]
> > Would love to have this as well. We use VS Code mostly for Angular work and we typically have multiple components (TS, HTML, CSS) and services (TS) open at a time, and we like to group relevant tabs together at the top.
> 
> This would be so useful! Ability to select multiple tabs and group them under single tab. And ideally option to have VSCode automatically group files with the same name but different extensions (eg test.component.ts/test.component.html/test.component.scss) under the same tab.

Could not agree more. Top row could be project level configs/settings, second row could be middle tier files, bottom row any front end files actively being worked on. This is how I imagine my workflow would look given the ability to do such a thing :)

**Comment by jreich1-godaddy** [comment]
Just tried this out in 1.53.0-insider. Works beautifully. Thank you!!!

**Comment by yulei900609** [comment]
When will this feature be merged into formal release?

**Comment by gjsjohnmurray** [comment]
> When will this feature be merged into formal release?

Expect your 1.52.1 to update to 1.53 later this week.

**Comment by mocsy** [comment]
I just got 1.53.0
From the change log: "A new setting workbench.editor.wrapTabs lets editor tabs wrap instead of showing a scrollbar."
Yay!

**Comment by fraigo** [comment]
Thanks to everyone working to push this out, specially @SneakyFish5 and @bpasero for the hard work!
![image](https://user-images.githubusercontent.com/8151943/107269555-9f4a7a00-69fe-11eb-8e68-d332d4ed4691.png)

**Comment by EliasCK** [comment]
Now, can we get some love for the fokes who wants to hide the tab bar area completely :) ? [https://github.com/microsoft/vscode/issues/33607](https://github.com/microsoft/vscode/issues/33607)
